### ISSUE TITLE
List hero IDs, ability IDs, and player item IDs throughout replay.

### ISSUE THREAD

**Issue by austinmilt** [skadistats/clarity#126]
I am a novice Java programmer looking to print the names/IDs of heroes, their abilities, and player items at various times throughout a replay. Clarity is obviously the right choice, but having spent the better part of a day trying to orient myself, I am very lost. Could someone give me a first step (other than just pointing to the existing examples) on how to retrieve this information?

In case it helps, I have gotten as far as being able to run various examples in clarity-examples for a given replay, as well as use the clarity-analyzer. Inspecting the .java source files is not really helping me because there's little indication of how to interface with game entities. 

Looking at the clarity analyzer, I see I want to spit out CDOTA_Unit_Hero_*, CDOTA_Unit_Hero_*.m_hAbilities.00*, and CDOTA_Unit_Hero_*.m_hItems.00* along with timestamps (or alternatively specific time requests) for the latter two.

Thanks for any help you're willing to provide.

**Comment by spheenik** [comment]
To get at entity data, you have to annotate your processor with `@UsesEntities` to tell clarity to parse entities. Then you can get a reference to the entity processor at any time by using `e = ctx.getProcessor(Entities.class)` (ctx being the parsing context object). From there you can query the current entities, and retrieve the wanted properties from them. 

Examples to see this are [the matchend example](https://github.com/skadistats/clarity-examples/blob/master/src/main/java/skadistats/clarity/examples/matchend/Main.java), and the [parser opendota uses](https://github.com/odota/parser/blob/master/src/main/java/opendota/Parse.java). Good luck!

**Comment by austinmilt** [comment]
Great, thanks for your help! I'll start there and see how it goes.

**Comment by chard234** [comment]
It looks like I am trying to do exactly the same thing!  I have managed to get a handle to the item entities, however I am now trying to identify what the item is.  There doesn't seem to be an obvious property allowing me to do this.  I am guessing the best way to find the name of the entity is using "m_pEntity.m_nameStringableIndex" but the value held here doesn't seem to line up with anything in the StringTables CombatLogNames.  Could you please give me some guidance on identifying the name/id of a CDOTA_Item.

@austinmilt I am happy to send you my source if it helps.

Thanks for the awesome parser :D

**Comment by spheenik** [comment]
Wearable items have a property `CBodyComponent.m_hModel`. It a so called resource handle.
You can use this handle by annotating with `@UsesResources`, and using the Resource processor with `ctx.getProcessor(Resources.class).getEntryForResourceHandle(*yourHandle*)`.

CDOTA_Item has `m_pEntity.m_nameStringableIndex` which is an index into the String table "EntityNames": Annotate using `@UsesStringTable("EntityNames")`, and get `ctx.getProcessor(StringTables.class).forName("EntityNames").getNameByIndex(*yourIndex*)`.

Cheers!

**Comment by austinmilt** [comment]
@chard234 I would love to get your source code as a starting point for me. I haven't had time to work on this again until today, and a jumpstart would be great! I'd suggest posting it complete here to help future lost souls.

**Comment by austinmilt** [comment]
Time for an update and some questions. Below I have code that does most of what I need, but not all of what I want. It accomplishes the following:

- Reads the list of hero names as npc_dota_hero_<name>
- For a known hero name from above, gets the hero's ability IDs and item IDs (ints) at each tick of the match
- Writes the results of the replay log to a given file.

Currently I'm lacking some important functionality and have questions as to how to accomplish that functionality.

1. What is the most robust means, given a .dem, to convert between the npc_dota_hero_<name> format and the CDOTA_Unit_Hero_<name> format? Is there info embedded in the .dem that will make this simple, or do I need to pull info from the game files or some other resource (c.f. https://github.com/skadistats/skadi/wiki/Hero-Names-and-IDs)?

2. Alternatively (and preferably), how can I directly extract the CDOTA_Unit_Hero_<name> hero names at each tick in the .dem?

3. Similarly, what is the most robust way to convert hero ability and item IDs to their common names or something other than an integer?

4. How do I get the current game time at each tick? I see this appeared in the old clarity.match, but I dont see it in clarity now.

5. Alternatively/in addition, what is the most efficient way to have the runner only process the game every X-seconds of game time?

6. Currently I'm editing the matchend example in-place. I've never used Maven before this, so I dont understand the "correct" way to build my package (see below for my method thus far). How do I customize my package and simply include clarity as a dependency? My attempts to do this have failed thus far.

My custom code in clarity-examples-master\src\main\java\skadistats\clarity\examples\matchend\Main.java
`package skadistats.clarity.examples.matchend;

import skadistats.clarity.Clarity;
import skadistats.clarity.decoder.Util;
import skadistats.clarity.model.Entity;
import skadistats.clarity.processor.runner.Context;
import skadistats.clarity.processor.entities.Entities;
import skadistats.clarity.processor.entities.UsesEntities;
import skadistats.clarity.processor.runner.SimpleRunner;
import skadistats.clarity.processor.reader.OnTickStart;
import skadistats.clarity.source.MappedFileSource;
import skadistats.clarity.wire.common.proto.Demo.CDemoFileInfo;
import skadistats.clarity.wire.common.proto.Demo.CGameInfo.CDotaGameInfo;
import skadistats.clarity.wire.common.proto.Demo.CGameInfo.CDotaGameInfo.CPlayerInfo;

import java.io.IOException;
import java.io.BufferedWriter;
import java.io.OutputStreamWriter;
import java.io.FileOutputStream;
import java.util.Arrays;

public class Main {

    // CONSTANTS
    public static final int MAX_PLAYERS = 50;
    public static final int ABILITY_SLOTS = 24;
    public static final int ITEM_SLOTS = 17;
    public static final String ITEM_SLOT_FMT = "m_hItems.%04d";
    public static final String ABILITY_SLOT_FMT = "m_hAbilities.%04d";
    public static final String CSV_SEP_ARRAY = ";";
    public static final String CSV_SEP_CAT = ",";
    
    public static String[] heroesNames = new String[MAX_PLAYERS];
    public static int nPlayers;
    private final SimpleRunner runner;
    private final BufferedWriter writer;
    
    public String getItemSlotNameByInt(int id) {
        return String.format(ITEM_SLOT_FMT, id);
    }
    
    public String getAbilitySlotNameByInt(int id) {
        return String.format(ABILITY_SLOT_FMT, id);
    }
        
    @UsesEntities
    @OnTickStart
    public void onTickStart(Context ctx, boolean synthetic) throws IOException {
        
        // basic declarations
        String heroName;
        Entity hero;
        String abilitySlotName;
        String itemSlotName;
        String heroLine;
        int[] abilityIDs = new int[ABILITY_SLOTS];
        int[] itemIDs = new int[ITEM_SLOTS];
        int j;
        
        Entities entities = ctx.getProcessor(Entities.class);
        
        // process the current abilities and items of every hero in this match
        for (int i = 0; i < nPlayers; i++) {
        
            // get the hero Entity from its DTClassName
            heroName = heroesNames[i];
            if (heroName.equals("npc_dota_hero_weaver")) {
                hero = entities.getByDtName("CDOTA_Unit_Hero_Weaver");
            }
            
            else if (heroName.equals("npc_dota_hero_drow_ranger")) {
                hero = entities.getByDtName("CDOTA_Unit_Hero_DrowRanger");
            }
            
            else {
                hero = null;
            }
            
            // get hero abilities and items (if hero has been set yet)
            if (hero != null) {
            
                // get IDs of hero abilities
                heroLine = heroName + CSV_SEP_CAT;
                for (j = 0; j < ABILITY_SLOTS; j++ ){
                    abilitySlotName = getAbilitySlotNameByInt(j);
                    abilityIDs[j] = hero.getProperty(abilitySlotName);
                    
                    // build the line to write to the log
                    if (j > 0) {
                        heroLine += CSV_SEP_ARRAY + Integer.toString(abilityIDs[j]);
                    }
                    
                    else {
                        heroLine += Integer.toString(abilityIDs[j]);
                    }
                }
                
                // get IDs of hero items
                heroLine += CSV_SEP_CAT;
                for (j = 0; j < ITEM_SLOTS; j++ ){
                    itemSlotName = getItemSlotNameByInt(j);
                    itemIDs[j] = hero.getProperty(itemSlotName);
                    
                    // build the line to write to the log
                    if (j > 0) {
                        heroLine += CSV_SEP_ARRAY + Integer.toString(itemIDs[j]);
                    }
                    
                    else {
                        heroLine += Integer.toString(itemIDs[j]);
                    }
                }
                
                // write to the log
                writer.write(heroLine + "\n");
            }  
        }
    }
    
    public Main(String inputFile, String outputFile) throws IOException, InterruptedException {
        writer = new BufferedWriter(
            new OutputStreamWriter(
                new FileOutputStream(outputFile), "utf-8"
            )
        );
        runner = new SimpleRunner(new MappedFileSource(inputFile)).runWith(this);
        writer.close();
    }
    
    public static void main(String[] args) throws Exception {
        
        // load final hero list by grabbing match info from the file using
        //  clarity's infoForFile
        CDotaGameInfo dotaInfo = new Clarity().infoForFile(args[0]).getGameInfo().getDota();
        nPlayers = dotaInfo.getPlayerInfoCount();
        CPlayerInfo player;
        for (int i = 0; i < nPlayers; i++) {
            player = dotaInfo.getPlayerInfo(i);
            heroesNames[i] = player.getHeroName();
        }
        
        // create the match analyzer (runner), which will search for item and
        //  ability changes as the match progresses
        new Main(args[0], args[1]);
    }
}`

To get this working, I do the following (in a terminal in Windows):
`cd C:\path\to\clarity-examples-master
mvn -P matchend package
java -jar target\matchend.one-jar.jar path\to\replay.dem log.csv
`

As one final note, I realized at some point today that using Maven to compile clarity _does_ actually create a simple javadoc in clarity-master\target\apidocs, which helps tremendously. It would be much better if any of the classes had documentation strings, but at least this gives a way to work through the structure of the code (see #127).

**Comment by spheenik** [comment]
Some answers:

1. Same as items: Entities have `m_pEntity.m_nameStringableIndex` (sic) which is an index into the String table "EntityNames": Annotate using `@UsesStringTable("EntityNames")`, and get `ctx.getProcessor(StringTables.class).forName("EntityNames").getNameByIndex(*yourIndex*)`.

2. If you have an entity, you can get it's name by `e.getDtClass().getName()`. There is a function on Entities.class to get all available entities which match a predicate (called `getAllByPredicate()`). If you supply a predicate matching the entity name like so, you have them all.

3. It should be possible to do `entities.getByHandle(*handle*)` without NPE. The items are in `m_hItems.00XX` and the abilities in `m_hAbilities.00XX`. There's a special handle 16777215, which means "empty". If you get the NPE from this, I'd be wanting to fix this in clarity.

4. You get the NPE because the GameRulesProxy is not there on early ticks. You have to do a null check and not work on it.

5. If you use 2.2, there's a [nifty feature](https://github.com/skadistats/clarity-examples/blob/master/src/main/java/skadistats/clarity/examples/propertychange/Main.java) to get called on property changes.

Ok, gtg now, answer for 6 has to wait.

**Comment by spheenik** [comment]
5. in addition: a tick is 1/30 of a second. So if you want to process only once per second, you can also exit your onTickStart() when the last time you processed is less than 30 ticks ago.

6. Are you using an IDE?

**Comment by chard234** [comment]
@austinmilt i am building a very similar data-set, it would be interesting for us to chat, and easier for me to help then over github.  

I created a temporary email address,  max@sharklasers.com (i'll delete this comment in a day or two as i am fairly sure posting emails is not good etiquette, but i can't think of a better way to exchange contact info) shoot me an email if you want to chat.

**Comment by austinmilt** [comment]
@spheenik Thank you for your continued help! I think I pretty much have a finished product here (at least for version 1). 

Below is code that runs through a replay, looking for updates to hero items and abilities. When these occur, it reports the hero's current items and abilities along with a timestamp of the real game time in the match (e.g. as would be reflected by the in-game clock at the top of the screen). Results are written to a CSV-style log, with each line being an update to a hero's items or abilities.

@spheenik I used your advice in several ways, my findings/decisions here:

1., 2., and 5. I decided to use the `@OnEntityPropertyChanged` decorator to avoid looking at every tick in the game needlessly, since I care only about changes to the status of heroes' items and abilities. With that, I was able to just have a hero Entity passed to my update method, and could get the hero name more easily from the StringTables using that. Note that because of what I decided to do with my number 3 question (see below), I decided to stick with the npc_dota_hero_... naming format (i.e. what's in the StringTables).

3. Despite being able to use entities.getByHandle(abilityIDs[j]).getDtClass().getDtName() when the item or entity was not null (thus avoiding NPE), I found that often the name of an item's or ability's DTClass is often generic, e.g. CDOTA_Item or some such. This is despite the fact that the item or ability isnt generic, e.g. hero talents received this designation even though their handles are unique and have different names. I don't know whether this is intended behavior (i.e. that they all have the same DTClass), but it was not what I needed. However, each item/ability in the StringTables (gotten from their StringTable index as explained in @spheenik's answer  to my number 1 question) does have a human-readable name, and so I decided to use that instead. Consequently, I created a method to get any entity's name from the StringTables, and so decided to use this same convention for hero names.

4. I added various null checks and it all works out. In fact, I think I made a pretty neat method for getting the real game time at a tick in a match (see the getRealGameTimeSeconds() method and getClockTime() method).

6. No, I am not using an IDE. However, I am able to get functioning compiled code with some changes to the directory structure and pom.xml (detailed below). I am still running into an issue when I compile this in its own directory versus as the matchend example. I get a whole bunch of DEBUG messages from clarity during the execution of my one-jar, e.g. `13:39:48.839 [main] DEBUG clarity.entities - entity at index 516 was not found when ordered to delete.` among many others. These do not occur when I use the one-jar from the same code replacing the matchend example. How do I turn these DEBUG messages off?

OK, below is my code and below that the commands to get it compiled using maven.

Main.java
```
package my_processor;

import skadistats.clarity.Clarity;
import skadistats.clarity.decoder.Util;
import skadistats.clarity.model.Entity;
import skadistats.clarity.model.DTClass;
import skadistats.clarity.model.FieldPath;
import skadistats.clarity.model.StringTable;
import skadistats.clarity.processor.entities.Entities;
import skadistats.clarity.processor.entities.OnEntityPropertyChanged;
import skadistats.clarity.processor.entities.UsesEntities;
import skadistats.clarity.processor.reader.OnTickStart;
import skadistats.clarity.processor.runner.Context;
import skadistats.clarity.processor.runner.SimpleRunner;
import skadistats.clarity.processor.stringtables.StringTables;
import skadistats.clarity.processor.stringtables.UsesStringTable;
import skadistats.clarity.source.MappedFileSource;
import skadistats.clarity.util.Predicate;
import skadistats.clarity.wire.common.proto.Demo.CDemoFileInfo;
import skadistats.clarity.wire.common.proto.Demo.CGameInfo.CDotaGameInfo;
import skadistats.clarity.wire.common.proto.Demo.CGameInfo.CDotaGameInfo.CPlayerInfo;

import java.io.IOException;
import java.io.BufferedWriter;
import java.io.OutputStreamWriter;
import java.io.FileOutputStream;
import java.util.Arrays;
import java.util.Iterator;

public class Main {

    // CLASS VARIABLES ***************************************************** //
    
    // constants
    public static final String ABL_PATTERN = "m_hAbilities.*";
    public static final int ABILITY_SLOTS = 24;
    public static final String ABILITY_SLOT_FMT = "m_hAbilities.%04d";
    public static final String CSV_ENCODE = "utf-8";
    public static final String CSV_SEP_ARRAY = ";";
    public static final String CSV_SEP_CAT = ",";
    public static final String CSV_NEWLINE = "\n";
    public static final String ENTITY_INDEX_PROP = "m_pEntity.m_nameStringableIndex";
    public static final String GRP_STR = "CDOTAGamerulesProxy";
    public static final String HERO_PATTERN = "CDOTA_Unit_Hero_.*";
    public static final String ITEM_PATTERN = "m_hItems.*";
    public static final int ITEM_SLOTS = 17;
    public static final String ITEM_SLOT_FMT = "m_hItems.%04d";
    public static final int MAX_PLAYERS = 50;
    public static final String STRTABLE_ENTNAME = "EntityNames";
    public static final String TIME_CUR_PROP = "m_pGameRules.m_fGameTime";
    public static final Float TIME_EPS = new Float(0.01);
    public static final String TIME_FMT = "%.3f";
    public static final String TIME_PRE_PROP = "m_pGameRules.m_flPreGameStartTime";
    public static final String TIME_SEP = ":";
    public static final String TIME_STA_PROP = "m_pGameRules.m_flGameStartTime";
    public static final String TIME_TRN_PROP = "m_pGameRules.m_flStateTransitionTime";
    
    // class globals
    // public static int nPlayers;
    private final SimpleRunner runner;
    private final BufferedWriter writer;
    private Float lastTime = new Float(-9999.0);
    private String lastHero = "zippy the frog";
    
    
    
    // METHODS ************************************************************* //
    
    /**
    * Creates the string representation of a hero's item slot property.
    * <p>
    * This should be used to automatically get the formatted name of the item
    * slot property for a hero entity in a replay file, to be used along with
    * the Entity.getProperty() method.
    *
    * @param id  the index ID of the item slot to get the property name for
    * @return the property name of the given item slot index
    */
    public String getItemSlotNameByInt(int id) {
        return String.format(ITEM_SLOT_FMT, id);
    }
    
    
    /**
    * Creates the string representation of a hero's ability slot property.
    * <p>
    * This should be used to automatically get the formatted name of the
    * ability slot property for a hero entity in a replay file, to be used 
    * along with the Entity.getProperty() method.
    *
    * @param id  the index ID of the ability slot to get the property name for
    * @return the property name of the given ability slot index
    */
    public String getAbilitySlotNameByInt(int id) {
        return String.format(ABILITY_SLOT_FMT, id);
    }
    
    
    /**
    * Gets the name of an entity from a replay's {@code StringTables}.
    * <p>
    * This should be used to get a human-readable tag for an {@code Entity} (as
    * opposed to a handle or other representation) as represented in a replay's
    * {@code StringTables}. It is intended to be used during the execution of a
    * runner (e.g. {@code SimpleRunner}) that {@code UsesStringTables}.
    *
    * @param id  the handle of the {@code Entity} for which a name should be retrieved
    * @param entities  the entities for the current tick of the current replay (e.g. as returned by Context.getProcessor(Entities.class))
    * @param stringTable  the replay {@code StringTable} from which to grab the entity's name
    * @return the name of the entity from [stringTable], null if the entity is null
    */
    public String getEntityNameByHandle(int id, Entities entities, StringTable stringTable) {
        Entity entity = entities.getByHandle(id);
        int index;
        if (entity == null) {
            return null;
        }
        else {
            index = entity.getProperty(ENTITY_INDEX_PROP);
            return stringTable.getNameByIndex(index);
        }
    }
    
    /**
    * Gets the real game time of a match in seconds as reflected in the in-game clock.
    * <p>
    * This gets the game time at the current tick in the runner of a replay 
    * from the {@code Entities}. It is intended to be used during the execution
    * of a runner (e.g. {@code SimpleRunner}).
    *
    * @param entities  the entities for the current tick of the current replay (e.g. as returned by Context.getProcessor(Entities.class))
    * @return the current game time in seconds, null before pre-game begins, negative during pre-game, positive after 0:00 of game clock
    */
    public Float getRealGameTimeSeconds(Entities entities) {
        Entity grules = entities.getByDtName(GRP_STR);
        Float gameTime = null;
        Float startTime = null;
        Float preGameTime = null;
        Float transitionTime = null;
        Float realTime = null;
        
        // before the match starts, there's CDOTAGamerulesProxy
        if (grules != null) {
            gameTime = grules.getProperty(TIME_CUR_PROP);
            
            // before the match starts, there's no game "time"
            if (gameTime != null) {
                preGameTime = grules.getProperty(TIME_PRE_PROP);
                
                // before hero picking and strategy time are finished, the
                //  pre-game countdown is still at 0, i.e. nothing has happened
                //  in the match
                if (preGameTime > TIME_EPS) {
                    startTime = grules.getProperty(TIME_STA_PROP);
                    
                    // time after the clock hits 0:00
                    if (startTime > TIME_EPS) {
                        realTime = gameTime - startTime;
                    }
                    
                    // between the pre-game and 0:00 time of the match, the
                    //  transition time reflects when the match is supposed to
                    //  start (i.e. hit 0:00 on the clock), and gives a good
                    //  approximation of when the match will start. Up to that
                    //  point, the start time is set to 0.
                    else {
                        transitionTime = grules.getProperty(TIME_TRN_PROP);
                        realTime = gameTime - transitionTime;
                    }
                }
            }
        }
        
        return realTime;
    }
    
    // NOT USED, BUT HERE FOR POSTERITY ***************************************
    // public String getClockTime(Entities entities) {
        // Float gameTime = getRealGameTimeSeconds(entities);
        // String clockTime = null;
        // if (gameTime != null) {
            // int minutes = (int) Math.floor(gameTime / 60.);
            // int seconds = (int) Math.round(Math.abs(gameTime % 60.));
            // clockTime = String.format("%d%s%02d", minutes, TIME_SEP, seconds);
        // }
        // return clockTime;
    // }
    // ************************************************************************
    
    @UsesEntities
    @UsesStringTable(STRTABLE_ENTNAME)
    /**
    * Writes a hero's attributes and items to file at a runner's current tick.
    * <p>
    * This is the most important method of this class, processing the abilities
    * and items of a hero at the runner's current tick. It writes the hero's
    * current abilities and items to an output file with a time-stamp of the
    * real game time of the current tick, in the format
    *
    * 5.555,hero_a,ability_1;ability_2;...;ability_n,item_1;item_2...;item_n
    * 7.344,hero_b,ability_1;ability_2;...;ability_n,item_1;item_2...;item_n
    *
    * Note how different separators are used for categories of things to be
    * reported for a hero on each line of the file.
    * <p>
    * Note this attempts to avoid duplication of hero status updates at high
    * frequency by checking for recent updates to heroes.
    *
    * @param ctx  the {@code Context} of the current runner
    * @param hero  the hero {@code Entity} for which to get updates
    * @throws IOException passed from other code that throws the same
    */
    public void heroUpdate(Context ctx, Entity hero) throws IOException {
        
        // loop declarations
        String abilitySlotName;
        int abilityID;
        Entity ability;
        int abilityIndex;
        String abilityName;
        
        String itemSlotName;
        int itemID;
        Entity item;
        int itemIndex;
        String itemName;
        
        int j;
        
        Entities entities = ctx.getProcessor(Entities.class);
        StringTable stringTable = ctx.getProcessor(StringTables.class).forName(STRTABLE_ENTNAME);
        
        // get current game time
        Float gameTime = getRealGameTimeSeconds(entities);
        
        // get hero abilities and items (if hero has been set yet)
        if (hero != null) {
        
            // get the hero name and see if this is another update in the same 
            //  time. If so, skip it
            String heroName = getEntityNameByHandle(hero.getHandle(), entities, stringTable);
            if (Math.abs(gameTime - lastTime) < TIME_EPS) {
                lastTime = gameTime;
                if (heroName.equals(lastHero)) {
                    return;
                }
                else {
                    lastHero = heroName;
                }
            }
            lastTime = gameTime;
            
            // get hero abilities
            String heroLine = String.format(TIME_FMT, gameTime) + CSV_SEP_CAT + heroName + CSV_SEP_CAT;
            for (j = 0; j < ABILITY_SLOTS; j++ ){
                abilitySlotName = getAbilitySlotNameByInt(j);
                abilityID = hero.getProperty(abilitySlotName);
                abilityName = getEntityNameByHandle(abilityID, entities, stringTable);
                
                // build the line to write to the log
                if (j > 0) {
                    heroLine += CSV_SEP_ARRAY + abilityName;
                }
                
                else {
                    heroLine += abilityName;
                }
            }
            
            // get hero items
            heroLine += CSV_SEP_CAT;
            for (j = 0; j < ITEM_SLOTS; j++ ){
                itemSlotName = getItemSlotNameByInt(j);
                itemID = hero.getProperty(itemSlotName);
                itemName = getEntityNameByHandle(itemID, entities, stringTable);
                
                // build the line to write to the log
                if (j > 0) {
                    heroLine += CSV_SEP_ARRAY + itemName;
                }
                
                else {
                    heroLine += itemName;
                }
            }
            
            // write to the log
            writer.write(heroLine + CSV_NEWLINE);
        }  
    }
    
    
    @OnEntityPropertyChanged(classPattern = HERO_PATTERN, propertyPattern = ITEM_PATTERN)
    /**
    * Uses the OnEntityPropertyChanged decorator to automatically update hero status during a replay.
    * <p>
    * This is essentially a listener for changes to hero items during a replay.
    * It uses such events to update hero items and abilities.
    *
    * @param ctx  the {@code Context} of the current runner
    * @param hero  the hero {@code Entity} for which to get updates
    * @param fp  not used, present by necessity with the {@code OnEntityPropertyChanged} decorator
    * @throws IOException passed from other code that throws the same
    */
    public void onItemChange(Context ctx, Entity hero, FieldPath fp) throws IOException {
        heroUpdate(ctx, hero);
    }
    
    
    @OnEntityPropertyChanged(classPattern = HERO_PATTERN, propertyPattern = ABL_PATTERN)
    /**
    * Uses the OnEntityPropertyChanged decorator to automatically update hero status during a replay.
    * <p>
    * This is essentially a listener for changes to hero abilities during a replay.
    * It uses such events to update hero items and abilities.
    *
    * @param ctx  the {@code Context} of the current runner
    * @param hero  the hero {@code Entity} for which to get updates
    * @param fp  not used, present by necessity with the {@code OnEntityPropertyChanged} decorator
    * @throws IOException passed from other code that throws the same
    */
    public void onAbilityChange(Context ctx, Entity hero, FieldPath fp) throws IOException {
        heroUpdate(ctx, hero);
    }
    
    
    /**
    * Creates and executes a replay runner as well as the writer for abilities and items.
    * <p>
    *
    * @param inputFile  the path to the input replay file
    * @param outputFile  the path to the output file where hero abilities and items should be recorded
    * @throws IOException passed from other code that throws the same
    * @throws InterruptedException passed from other code that throws the same
    */
    public Main(String inputFile, String outputFile) throws IOException, InterruptedException {
        writer = new BufferedWriter(
            new OutputStreamWriter(
                new FileOutputStream(outputFile), CSV_ENCODE
            )
        );
        runner = new SimpleRunner(new MappedFileSource(inputFile)).runWith(this);
        writer.close();
    }
    
    
    /**
    * Creates and executes the Main class.
    * <p>
    *
    * @param args  the command-line args for {@code Main}
    * @throws Exception passed from other code that throws the same
    */
    public static void main(String[] args) throws Exception {
        
        // ***** NO LONGER USING BUT MAY USE LATER TO GET OTHER PLAYER INFO ***
        // // load final hero list by grabbing match info from the file using
        // //  clarity's infoForFile
        // CDotaGameInfo dotaInfo = new Clarity().infoForFile(args[0]).getGameInfo().getDota();
        // nPlayers = dotaInfo.getPlayerInfoCount();
        // CPlayerInfo player;
        // for (int i = 0; i < nPlayers; i++) {
            // player = dotaInfo.getPlayerInfo(i);
            // heroesNames[i] = player.getHeroName();
        // }
        // ********************************************************************
        
        // create the match analyzer (runner), which will search for item and
        //  ability changes as the match progresses
        new Main(args[0], args[1]);
    }
}
```

pom.xml
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>my_processor</groupId>
    <artifactId>my_processor</artifactId>
    <version>0.0</version>
    <packaging>jar</packaging>
    <name>my-processor</name>

    <properties>
        <maven.compiler.target>1.7</maven.compiler.target>
        <maven.compiler.source>1.7</maven.compiler.source>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <developers>
        <developer>
            <id>austinmilt</id>
            <name>Austin Milt</name>
            <email>austin.w.milt@gmail.com</email>
        </developer>
    </developers>

    <repositories>
        <repository>
            <id>sonatype.oss.snapshots</id>
            <name>Sonatype OSS Snapshot Repository</name>
            <url>http://oss.sonatype.org/content/repositories/snapshots</url>
            <releases>
                <enabled>false</enabled>
            </releases>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
        </repository>
    </repositories>

    <dependencies>
        <dependency>
            <groupId>com.skadistats</groupId>
            <artifactId>clarity</artifactId>
            <version>2.2-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>joda-time</groupId>
            <artifactId>joda-time</artifactId>
            <version>2.7</version>
        </dependency>
        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-classic</artifactId>
            <version>1.1.3</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>2.6</version>
                <configuration>
                    <archive>
                        <manifest>
                            <mainClass>my_processor.Main</mainClass>
                        </manifest>
                    </archive>
                </configuration>
            </plugin>
            <plugin>
                <groupId>com.jolira</groupId>
                <artifactId>onejar-maven-plugin</artifactId>
                <version>1.4.4</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>one-jar</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-javadoc-plugin</artifactId>
                <executions>
                    <execution>
                        <id>attach-javadocs</id>
                        <goals>
                            <goal>jar</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>

    <profiles>
        <profile>
            <id>my_processor</id>
            <properties>
                <exampleName>my_processor</exampleName>
            </properties>
        </profile>
    </profiles>

</project>
```

Directory structure is
my_processor-master\src\main\java\my_processor\Main.java
my_processor-master\pom.xml

To compile, in a Windows terminal
```
cd C:\path\to\my_processor-master
mvn -P my_processor package
```
Note this should produce a javadoc html documentation in target\apidocs\index.html.

And to run for a replay (dota 7.01)
```
java -jar target\my_processor-0.0.one-jar.jar C:\path\to\replay.dem log.txt
```

**Comment by spheenik** [comment]
Looks nice and clean :)

**Comment by austinmilt** [comment]
I'm back!

I've noticed that, for early times in my match, my code thinks that heroes have a null state for some ability slots, when I can clearly see that's not the case for those times in the clarity analyzer. For instance, here's one line from my output file

-74.867,npc_dota_hero_lina,null;null;null;null;null;null;null;null;null;special_bonus_attack_range_150;special_bonus_unique_lina_1;special_bonus_unique_lina_2;null;null;null;null;null;null;null;null;null;null;null;null,item_tango;item_wind_lace;item_clarity;null;item_ward_observer;null;null;null;null;null;null;null;null;null;null;null;null

The first element is the game time (in seconds), which is after the heroes have spawned into the map. The second is the hero name. The series of nulls are ability names in ability slots (e.g. the first null is the name of lina's m_hAbilities.0000 taken from the stringtables, i.e. this is ability ID 16777215). When I look at this time in the clarity analyzer, I see non-null abilities in the ability slots.

So, what's going on? Is this a symptom of using @OnEntityPropertyChanged?

**Comment by austinmilt** [comment]
I'm back!

I'm having an issue where early in the replay entities.getByHandle is returning null for entities that shouldnt be null. In particular, I'm trying to get hero abilities by their handle stored in CDOTA_Unit_Hero_.*.m_hAbilities.*. It appears that this occurs when @OnEntityPropertyChanged is triggered in the first moments of the game as heroes spawn on the map. Does this mean the game hasnt actually populated the entities at this point? If so, is there a way for me to make sure I dont get a trigger of @OnEntityPropertyChanged until all entities have been created? If not, @OnEntityPropertyChanged is flawed for my purposes because I need the ability names at the earliest moment.

**Comment by spheenik** [comment]
Can you please open a new issue for this?

**Comment by spheenik** [comment]
maybe #134 is related?
